### PR TITLE
Footprints!

### LINKED ISSUE #140: Forky Readiness: Smears/Staining/Washing
This is an issue tracker for various parts of the Forky readiness milestone goal, for Forky to be release-ready.

If a PR accomplishes the goal of a specific issue as part of this milestone, please link it as part of the PR.

If you have questions or comments on this issue please feel free to make them here. Thank you!

### PR DESCRIPTION
**Pull request by YaraaraY**
## About the PR
This PR is an original re-write of [Funky Station PR #2711](https://github.com/funky-station/funky-station/pull/2711) with functional parity and additional features & fixes. This also fixes a small mistake in #107 that made most masks not stainable.

## Why / Balance
See original PR.

## Technical details
Compared to the original implementation:
- Footprint visuals correctly update when cleaned using space cleaner
- Footprint offsets and rotation angles are relative to the actual movement trajectory
- Modified the UpdateSlow logic so footprint entities do not apply movement speed penalties, regardless of reagent
- Integrated footprint exclusions into the OnStepInPuddle event hook to prevent the Stains system from making footprints disappear
- Added the NoFootprintsComponent which can be used on galoches & other footwear we don't want leaving footprints

## Media
image

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.
- [X] If I am porting something, I have done my best to respect the appropriate licenses associated with the presented changes.

## License
MIT

**Changelog**
:cl: AraiMaia
- add: When stepping or crawling through reagent puddles, you now leave behind a trail of footprints or stains!
- add: Galoshes now prevent you from leaving behind footprints.

### CODE REVIEW

File: Resources/Prototypes/Body/species_base.yml [hunk 1]
```diff
@@ -231,3 +231,4 @@
   - type: FireVisuals
     alternateState: Standing
   - type: StunVisuals
+  - type: FootprintOwner
```
Comment (Chapien) [inline Resources/Prototypes/Body/species_base.yml]: Do we want to mark changes? I noticed other modifications to upstream files are marked ie #Funky Change or something along those lines.

File: Resources/Prototypes/Entities/Clothing/Masks/base_clothingmask.yml [hunk 2]
```diff
@@ -12,6 +12,10 @@
   - type: StaticPrice
     price: 25
   - type: Appearance
+  - type: SolutionContainerManager
```
Comment (Chapien) [inline Resources/Prototypes/Entities/Clothing/Masks/base_clothingmask.yml]: Again, consider marking this as a Funky change

File: Resources/Prototypes/Entities/Clothing/Shoes/specific.yml [hunk 3]
```diff
@@ -117,6 +117,7 @@
   - type: NoSlip
   - type: SlowedOverSlippery
     slowdownModifier: 0.7
+  - type: NoFootprints
```
Comment (Chapien) [inline Resources/Prototypes/Entities/Clothing/Shoes/specific.yml]: Consider marking as a Funky change

File: Resources/Prototypes/Entities/Effects/puddle.yml [hunk 4]
```diff
@@ -216,3 +216,49 @@
   - type: Tag
     tags:
     - DNASolutionScannable
+
```
Comment (Chapien) [inline Resources/Prototypes/Entities/Effects/puddle.yml:219]: Consider marking as a Funky change

File: Resources/Prototypes/Entities/Objects/Specific/Janitorial/janitor.yml [hunk 5]
```diff
@@ -38,6 +38,8 @@
     size: Large
     sprite: Objects/Specific/Janitorial/mop.rsi
   - type: Absorbent
+    footprintCleaningRange: 0.45
+    maxCleanedFootprints: 25
     useAbsorberSolution: true
```
Comment (Chapien) [inline Resources/Prototypes/Entities/Objects/Specific/Janitorial/janitor.yml:41]: Consider marking as a Funky change

**Review by Chapien — CHANGES_REQUESTED** [review]
This looks great to me. I tested it and it works perfectly, though I admit I didn't test all of the edge cases. I do question the performance cost of having every footstep be an entity, but that's a bit of an architectural problem of upstream and not really our fault/fixable on our end. We could maybe have footprints evaporate after a long enough time has passed, but until we start noticing a performance drop there's no need to worry. 

I did suggest marking Funky changes to upstream files, but other than that I'm happy to approve this.

**Comment by YaraaraY** [comment]
comments added!

**Comment by Chapien** [comment]
With the necessary comments added, I'm satisfied with the merging this, though I will wait out of respect for the other reviewers.

**Review by Chapien — APPROVED** [review]
